### PR TITLE
8266371: Memory segment bound check fails because of small segment optimizations

### PR DESCRIPTION
**Pull request by mcimadamore**
The optimization for small memory segment bound check does not take into account numeric overflow. As a result, the bounds check does not detect big long values which, when casted back to int values appear to be within the segment's bounds.

The fix is to avoid the optimized bound check routine if the offset/length at which the segment is accessed is bigger than Integer.MAX_VALUE.

Benchmarks do not seem affected by this change (I'll post results in a separate comment).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266371](https://bugs.openjdk.java.net/browse/JDK-8266371): Memory segment bound check fails because of small segment optimizations

### Reviewers
 * [Athijegannathan Sundararajan](https://openjdk.java.net/census#sundar) (@sundararajana - Committer) ⚠️ Review applies to 6f9d37540075fad76117c7ec7f38273feaaaaa11
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - Committer) ⚠️ Review applies to 2cbf6f5ef817bcfb41e79d58537272ce5bdcca45

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/524/head:pull/524` \
`$ git checkout pull/524`

Update a local copy of the PR: \
`$ git checkout pull/524` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/524/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 524`

View PR using the GUI difftool: \
`$ git pr show -t 524`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/524.diff">https://git.openjdk.java.net/panama-foreign/pull/524.diff</a>

</details>

### CODE REVIEW

File: test/micro/org/openjdk/bench/jdk/incubator/foreign/LoopOverNonConstantHeap.java [hunk 1]
```diff
@@ -97,11 +97,6 @@ public void setup() {
         byteBuffer = ByteBuffer.wrap(base).order(ByteOrder.nativeOrder());
     }
 
-    @TearDown
```
Comment (mcimadamore) [inline test/micro/org/openjdk/bench/jdk/incubator/foreign/LoopOverNonConstantHeap.java:100]: This change fix an issue with the benchmark which was caused by the new ResourceScope API: heap segments are associated with the global scope, so they cannot be closed.

**Review by sundararajana — APPROVED** [review]
LGTM

**Review by ChrisHegarty — APPROVED** [review]

**Review by sundararajana — APPROVED** [review]
Additional changes are fine.

**Pull request closed without merging**